# Master's Thesis Repository

This repository contains scripts, parameters, and workflows from my master's thesis.

## Repository Structure

### `ITS`
Contains the workflow and parameters used for the **Internal Transcribed Spacer (ITS)** analysis, as well as post-analysis scripts and graph generation. This directory includes all necessary components to process ITS data from initial sequencing to the final visualization of results.

### `WGS`
Contains the workflow and parameters used for the **Whole Genome Sequencing (WGS)** analysis, along with post-analysis scripts and tools for generating graphical representations. This directory encompasses the entire pipeline for WGS data processing.

### `Database-setup`
Contains the scripts necessary for creating a custom database from a list of taxa for use with **Kraken2**. These scripts facilitate the customization of taxonomic databases, allowing for tailored and accurate taxonomic classification based on specific research needs.

### `Co-ocurrence_network`
Contains the scripts for creating objects used in the construction of co-occurrence networks, as well as the necessary files for subsequent analysis with **Cytoscape**. This directory provides tools for building and analyzing complex networks to understand the relationships and interactions between different taxa.

### `ITS_vs_WGS`
Contains the scripts used to describe metrics and generate graphs that assist in the comparison of sequencing methods. This directory includes tools for evaluating and visualizing the performance and results of ITS versus WGS methodologies, enabling a comprehensive comparison of their effectiveness.


## Prerequisites

Ensure that you have the necessary software and dependencies installed for each analysis pipeline.
